File: java/leetcode/minimum-distance-between-bst-nodes.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public static int min;
    public int minDiffInBST(TreeNode root) {
        min = Integer.MAX_VALUE;
        dfsMinDiff(root);
        return min;
    }

    private void dfsMinDiff (TreeNode root) {
        if (root.left != null) {
            min = Math.min(min, Math.abs(root.val - dfRight(root.left)));
            dfsMinDiff(root.left);
        }
        if (root.right != null) {
            min = Math.min(min, Math.abs(root.val - dfLeft(root.right)));
            dfsMinDiff(root.right);
        }
    }

    private int dfLeft (TreeNode root) {
        if (root.left != null) {
            return dfLeft(root.left);
        } else {
            return root.val;
        }
    }

    private int dfRight (TreeNode root) {
        if (root.right != null) {
            return dfRight(root.right);
        } else {
            return root.val;
        }
    }
}

File: javascript/leetcode/minimum-distance-between-bst-nodes.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var minDiffInBST = function(root) {
    var dfsMinDiff = function (root) {
        if (root.left !== null) {
            min = Math.min(min, Math.abs(root.val - dfRight(root.left)));
            dfsMinDiff(root.left);
        }
        if (root.right !== null) {
            min = Math.min(min, Math.abs(root.val - dfLeft(root.right)));
            dfsMinDiff(root.right);
        }
    };

    var dfLeft = function (root) {
        if (root.left !== null) {
            return dfLeft(root.left);
        } else {
            return root.val;
        }
    };

    var dfRight = function (root) {
        if (root.right !== null) {
            return dfRight(root.right);
        } else {
            return root.val;
        }
    };
    min = Number.MAX_SAFE_INTEGER;
    dfsMinDiff(root);
    return min;
};

